DynamoDB: Only set throughput for PROVISIONED billing mode

nregina opened the pull request:
When creating DynamoDb tables with PAY_PER_REQUEST billing mode, the AWS API throws an exception if throughput (read capacity and write capacity) is populated in the create table request.  This error condition does not occur when testing against the dynamodb-local:1.13.6 docker image, but it does when running against DynamoDb in AWS.

With this change, throughput is only populated if billing mode is set to PROVISIONED.

nregina commented:
This change was already covered by [DynamoDbTicketRegistryFacilitatorTests.java](https://github.com/apereo/cas/blob/6.4.x/support/cas-server-support-dynamodb-ticket-registry/src/test/java/org/apereo/cas/ticket/registry/DynamoDbTicketRegistryFacilitatorTests.java).  However, since the dynamodb-local:1.13.6 docker image doesn't exhibit the same behavior as the AWS API, the bug, introduced in #5107, was not apparent until that code was run in an actual AWS account.  Until the dynamodb-local docker image exhibits the same behavior, it will be difficult to test this without running against a real AWS account.  I've run this change in my AWS account with billing mode set to both PROVISIONED and PAY_PER_REQUEST and the error condition did not recur.  The error does still occur with CAS 6.4.3.  I've added further tests to confirm the table properties when creating with each billing mode.

mmoayyed commented:
- Is this something that is fixed in amazon/dynamodb-local:1.17.0? 
- Is this something that is taken care of in localstack/localstack? If not, have you had a chance to start a discussion with them? 
- Are you saying that if I look at the coverage report by running DynamoDbTicketRegistryFacilitatorTests, your change is already covered?
- Is there a way you could design a test to look at the table metadata and confirm the correct mode was used to create the schema?

Also, when posting PRs, please always start with the master branch, and then backport as necessary. The other way around usually isn't ideal.

nregina commented:
This isn't fixed in amazon/dynamodb-local:1.17.0 nor is it fixed in localstack/localstack:latest.  This can be tested with the AWS CLI command:

### AWS
```
aws dynamodb create-table --table-name test-create-table --attribute-definitions '[{"AttributeName":"id", "AttributeType":"S"}]' --key-schema '[{"AttributeName":"id", "KeyType":"HASH"}]' --billing-mode PAY_PER_REQUEST --provisioned-throughput '{ "ReadCapacityUnits": 10, "WriteCapacityUnits": 10 }'
```
output
```
An error occurred (ValidationException) when calling the CreateTable operation: One or more parameter values were invalid: Neither ReadCapacityUnits nor WriteCapacityUnits can be specified when BillingMode is PAY_PER_REQUEST
```

### amazon/dynamodb-local:1.17.0 or localstack/localstack:latest
```
aws --endpoint-url=http://localhost:[8000 | 4566] dynamodb create-table --table-name test-create-table --attribute-definitions '[{"AttributeName":"id", "AttributeType":"S"}]' --key-schema '[{"AttributeName":"id", "KeyType":"HASH"}]' --billing-mode PAY_PER_REQUEST --provisioned-throughput '{ "ReadCapacityUnits": 10, "WriteCapacityUnits": 10 }'
```
output
```
{
    "TableDescription": {
        "AttributeDefinitions": [
            {
                "AttributeName": "id",
                "AttributeType": "S"
            }
        ],
        "TableName": "test-create-table",
        "KeySchema": [
            {
                "AttributeName": "id",
                "KeyType": "HASH"
            }
        ],
        "TableStatus": "ACTIVE",
        "CreationDateTime": "2021-11-22T09:31:40.829000-05:00",
        "ProvisionedThroughput": {
            "ReadCapacityUnits": 10,
            "WriteCapacityUnits": 10
        },
        "TableSizeBytes": 0,
        "ItemCount": 0,
        "TableArn": "arn:aws:dynamodb:us-east-1:000000000000:table/test-create-table",
        "TableId": "be6e6143-5031-4709-9432-3a7aa740a84d",
        "BillingModeSummary": {
            "BillingMode": "PAY_PER_REQUEST",
            "LastUpdateToPayPerRequestDateTime": "2021-11-22T09:31:40.829000-05:00"
        }
    }
}
```

Opened issue [4982](https://github.com/localstack/localstack/issues/4982) with localstack/localstack

The test I added in #5107 would have failed if the dynamodb-local behavior was consistent with the AWS API and would succeed with this change.  Since this PR was rejected, I've added tests that examine the table metadata when creating with billingMode PROVISIONED as well as PAY_PER_REQUEST to verify that throughput is only set when the table is created in PROVISIONED mode.  That should cover this change, but again, it's only verifying what was set and not that it wouldn't potentially fail when run against a real AWS account.  I did, however, verify the behavior against my AWS account.

mmoayyed commented:
Thank you. I appreciate the thorough analysis. Have you pushed your recent changes/tests to the branch here? And could you please create a new PR targeted at master, or would you prefer I port forward manually and cherry pick?

nregina commented:
Great!  I've added the changes to the 6.4.x branch in my fork, but they're not showing in this PR.  I assume that's because the PR is closed.  Should I open a new PR or are you able to reopen this one?

I've already started porting it forward into master so I'll create a PR for that.  The create table code has been refactored a bit between 6.4.x and master so the changes aren't in the same classes.

mmoayyed commented:
Re-opening this again; hopefully the bot is kind to us all.
